hs-bindgen/src-internal/HsBindgen/C/Tc/Macro.hs

```diff
@@ -827,76 +821,120 @@ inferFun f@( Fun fun ) =
-  MDiv        -> q2 \ a b -> QuantTyBody [Div  a b] $ funTy [a,b] ( DivRes  a b )
-  MRem        -> q2 \ a b -> QuantTyBody [Rem  a b] $ funTy [a,b] ( RemRes  a b )
+  MMult       -> q2 \ a b -> QuantTyBody [Mult a b] ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singMult ty1 ty2), mkFunTy [a,b] ( MultRes a b ) )
+  MDiv        -> q2 \ a b -> QuantTyBody [Div  a b] ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singDiv  ty1 ty2), mkFunTy [a,b] ( DivRes  a b ) )
+  MRem        -> q2 \ a b -> QuantTyBody [Rem  a b] ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singRem  ty1 ty2),  mkFunTy [a,b] ( RemRes  a b ) )
 
-    -- Bit
-  MBitwiseNot -> q1 \ a   -> QuantTyBody [Complement a] $ funTy [a]   ( ComplementRes a )
-  MBitwiseAnd -> q2 \ a b -> QuantTyBody [Bitwise a b]  $ funTy [a,b] ( BitsRes a b )
-  MBitwiseXor -> q2 \ a b -> QuantTyBody [Bitwise a b]  $ funTy [a,b] ( BitsRes a b )
-  MBitwiseOr  -> q2 \ a b -> QuantTyBody [Bitwise a b]  $ funTy [a,b] ( BitsRes a b )
+    -- Bitwise logical operators
+  MBitwiseNot -> q1 \ a   -> QuantTyBody [Complement a] ( unaryFun  $ \ ty      f -> f (C.Expr.singComplement ty) , mkFunTy [a]   ( ComplementRes a ) )
+  MBitwiseAnd -> q2 \ a b -> QuantTyBody [Bitwise a b]  ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singBitAnd ty1 ty2), mkFunTy [a,b] ( BitsRes a b ) )
+  MBitwiseXor -> q2 \ a b -> QuantTyBody [Bitwise a b]  ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singBitXor ty1 ty2), mkFunTy [a,b] ( BitsRes a b ) )
+  MBitwiseOr  -> q2 \ a b -> QuantTyBody [Bitwise a b]  ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singBitOr  ty1 ty2), mkFunTy [a,b] ( BitsRes a b ) )
 
     -- Bit shift
-  MShiftLeft  -> q2 \ a i -> QuantTyBody [Shift a i] $ funTy [a,i] ( ShiftRes a )
-  MShiftRight -> q2 \ a i -> QuantTyBody [Shift a i] $ funTy [a,i] ( ShiftRes a )
+  MShiftLeft  -> q2 \ a i -> QuantTyBody [Shift a i] ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singShiftL ty1 ty2), mkFunTy [a,i] ( ShiftRes a ) )
+  MShiftRight -> q2 \ a i -> QuantTyBody [Shift a i] ( binaryFun $ \ ty1 ty2 f -> f (C.Expr.singShiftR ty1 ty2), mkFunTy [a,i] ( ShiftRes a ) )
   where
     q1 body = Quant @( S Z )       \ (a ::: VNil) -> body a
     q2 body = Quant @( S ( S Z ) ) \ (a ::: i ::: VNil) -> body a i
-    funTy mbArgs res =
-      case NE.nonEmpty mbArgs of
-        Just args -> FunTy args res
-        Nothing   -> res
+
+    unaryFun :: ( forall ty r. C.Type.SType ValSType ty
+                  -> ( forall res. ( C.Type.SType ValSType res, ty -> res ) -> r ) -> r )
+             -> FunValue
+    unaryFun proveFn =
+      FunValue @( S Z ) ( Text.pack ( show fun ) ) $ \ ( a ::: VNil ) ->
+        if | Value ( ValSType ty ) x <- a
+           -> proveFn ty $ \ ( resTy, fn ) ->
+                Value ( ValSType resTy ) ( fn x )
+           | otherwise
+           -> NoValue
+
+    binaryFun :: ( forall ty1 ty2 r. C.Type.SType ValSType ty1 -> C.Type.SType ValSType ty2
+                   -> ( forall res. ( C.Type.SType ValSType res, ty1 -> ty2 -> res ) -> r ) -> r )
```

Comment:
This is trying to quantify over the type family, e.g.

```hs
forall ty1 ty2 tf. C.Type.SType ValSType ty1 -> C.Type.SType ValSType ty2
  -> (C.Type.SType ValSType (tf ty1 ty2), ty1 -> ty2 -> tf ty1 ty2)
```

but that doesn't work because GHC interprets quantification over `tf` to be quantification over a matchable type constructor such as `Maybe`, and doesn't allow it to be instantiated with a (necessarily partially applied) type family.

Maybe we can try to move the quantification around:

```hs
    binaryFun :: forall res. ( forall ty1 ty2. C.Type.SType ValSType ty1
              -> C.Type.SType ValSType ty2
              -> (C.Type.SType ValSType res, ty1 -> ty2 -> res ) )
```

Reply:
This doesn't work because we would have to instantiate `res` to a type that mentions inner type variables `ty1`, `ty2`.